### ISSUE TITLE
Add an example of a calculation with nprocesses>1

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#272]
Just a reminder. In the epochX work #244, I kept several of the comments I had all over the code "FIXME this assumes nprocesses=1". I am not sure what the code should look like or how it should behave with nprocesses>1.

Eventually we need to add a test of one simple such example. This may even be a very simple process, as long as it has nprocesses>1.

@oliviermattelaer maybe this is for you to suggest one ;-)

Low priority anyway (Fortran integration is more important)

**Comment by valassi** [comment]
Adding pptt is definitely needed - Olivier's bug #337 probably comes from this

**Comment by oliviermattelaer** [comment]
So in the latest gpucpp branch.
All P directories would have exactly one matrix-element function.
This matrix-element can still be related to multiple initial/final state but all of them will have the exact same matrix-element.

(Compare to my original email, this is now done in a "decent" way meaning that identical matrix-element are correctly handle)

**Comment by valassi** [comment]
Thanks Olivier! I updated to your latest upstream code in PR #619.

This is related to the discussion we had at the meeting earlier this wek https://indico.cern.ch/event/1263518/
You showed that for instance this process changes as follows
![image](https://user-images.githubusercontent.com/3473550/229080358-ae23dccc-dcca-4b16-a9c4-c7cc4430bf65.png)

I could close this as fixed by PR #619. But I will keep it open for now, maybe I will clean up the code, remove a lot of nprocesses and/or add a few cleaner comments mentioning this #272 and PR #619

**Comment by valassi** [comment]
In the end I have done a lot more analysis about this issue. I did not want to simply merge MR #619 and close this one. I wanted to make sure that before the MR we had a problem with nprocesses>1, and after the MR we did not.

This took several iterations and extra issues opened on the way
- I wanted to reproduce Olivier's screesnhot above to see what has changed.
- I understood the above probably came from work done with Nathan on https://github.com/nscottnichols/madgraph4gpu/blob/f9689784390e2975a055711b3a8d79f7b7688860/epochX/sycl/CODEGEN/generateAndCompare2.sh#L58 on a SUSY example, so I tried several SUSY examples, initially gq_ttllq. On the way I opened #624 #625 #627 #628
- Then I actually realised I could probably just test a SM process like gq_ttq (but I also did gq_ttllq for SM), #626 (which also gave some of the issues above)
- Then I also tried in parallel to produce the plot with "launch", and this gave me #629 

Anyway, my understanding now is the following (I will discuss with Olivier)
- the issue we discuss here comes from things like gq_ttq where the initial and final q can be both u or a ubar
- within the fortran, the "nprocesses>1" means that there are actually DSIG1 and DSIG2 in a single P1 directory... correspondingly, one could have expected nprocesses=2 in cudacpp inside a single P1 directory
- however, one workaorund is to split this into two separate P1 directories: this is Olivier's split_nonidentical_grouping" and #619: I have checked that in the end this gives indeed two P1 subdirectories, with only one DSIG1 in each
- note well, Olivier's patch uses two P1 directpies ONLY if cudacpp is an exporter: so not in madevent with Fortran MEs... note also, the standalone cudacpp was ALREADY doing two separate directories... so the issue was really only madevent+cudacpp
- note that, by going to several P1 directories, we will hit even more the issues in #534 

So in a way this #272 (nprocesses>1) and the #534 (more than one P1) are related, because the sam eissues can be moved from the former to the latter

In practice now
- I think I will add gqttq with a single P1 and with DSIG1/DIG2 in #626
- on the way, in the above I will **CLEAN UP MESSAGES ABOUT NPROCESSES** in the code 
- then and only then I will merge #619, showing that  there are now two P1 with a single DSIG1
- then I will fix gq_ttq in this new configuration
Along the way, we need (in #536) also to add the infrastructure for tput/tmad tests (and builds!!) from more than one P1

**Comment by oliviermattelaer** [comment]
also it seems that some matrix element are missing (the symmetric ones) in the susy case, this need to be fix.

**Comment by valassi** [comment]
> also it seems that some matrix element are missing (the symmetric ones) in the susy case, this need to be fix.

thanks Olivier... can you elaborate please? which process? expect build/runtime failures or just wrong results? I am about to close this issue, we can open a new one (unless this is covered by the other issues already open for SUSY)

**Comment by valassi** [comment]
I am finally closing this. There have been a lot of changes (and I am not 100% sure I understand all the implications), but essentially cudacpp is now configured so that
- there are no "mirror processes" (which for instance were giving nprocesses=2 in uux_ttx in branch 270, see #343 and PRs #360 and #396)... which probably is what fixed this #272
- in addition (I think this is a different issue, even if also related to quark-antiquark exchange?) there are no processes with both a DSIG1 and DSIG2 (i.e. also both matrix1.f and matrix2.f), because of Olivier's "split_nonidentical_grouping" changes picked up in PR #619 (and tested in gq_ttq in PR #626, which also cleans up all comments about nprocesses==1)...  now those end up as two separate P1 directories instead

I have added some improved comments about the issues above in PR #635, which I have just merged.

One last word: having removed mirror processes and the possibility to have DSIG2, **now cudacpp relies on having several P1 subdirectories**. In some cases (actually not the pp_tt or gq_ttq processes above), there are problems because some files should be different in the various P1 and they are not: see for instance pp to tttt in WIP PR #534, where the value of nwf should be different. [NB I have tested that this issue happens in pp_tttt but not in pp_tt]

For the summary of the summary, about the "example processes"
- pp to tt (and specifically uux to ttx within that) was an example of mirror process, giving nprocesses>1 in branch 270 before #396, fixed long ago
- gq to ttq (or more precisily gu to ttu AND gux to ttux) was an example of a process with both DSIG1 and DSIG2 in the same P1 directory, before PRs #619 and #626, now fixed and resulting in several P1 subdirectories
- pp to tttt (see WIP PR #534) is an example of a process with many P1 subdirectoies, but where nwf must be different in the various directories (this is NOT The case of pp_tt or gq_ttq above, where all P1 builds are ok)
- **as a consequence, gq_ttq has been added to the repo and pp_tttt will also be added to the repo**

FINALLY closing this.

cc: @roiser @oliviermattelaer @zeniheisser
